### DISCUSSION TITLE
Profile-Guided Optimization (PGO) benchmark results

### DISCUSSION THREAD

**Discussion by zamazan4ik** [apache/datafusion#9507] [Ideas]
Hi!

I evaluate Profile-Guided Optimization (PGO) performance improvements for applications and libraries in different software domains - all current results can be found [here](https://github.com/zamazan4ik/awesome-pgo). According to the tests, PGO helps to improve performance in various software domains. I decided to perform PGO benchmarks on the `arrow-datafusion` library too since some library users may be interested in improving the library performance (e.g. @Dandandan suggested to test PGO on `arrow-datafusion` here: https://github.com/sqlparser-rs/sqlparser-rs/discussions/1163#discussioncomment-8658556)

I did some benchmarks and here are the results.

## Test environment

* Fedora 39
* Linux kernel 6.7.4
* AMD Ryzen 9 5900x
* 48 Gib RAM
* SSD Samsung 980 Pro 2 Tib
* Compiler - Rustc 1.76
* arrow-datafusion version: `main` branch on commit `767760b84c71b9af8b2d8cdee9cd5c099b57b462`

## Benchmark

For benchmark purposes, I use [these](https://github.com/apache/arrow-datafusion/tree/main/benchmarks) benchmarks. Since it consists of many scenarios, I ran each scenario with each dedicated training phase. It means that for the Clickbench benchmark only Clickbench is used during the PGO training phase, for TPCH - only TPCH, etc. I made such a choice since I don't know how different are these benchmarks internally (OLTP vs OLAP, maybe other huge differences).

Actual benchmark commands are taken from the [bench.sh](https://github.com/apache/arrow-datafusion/blob/main/benchmarks/bench.sh) with exact parameters like iteration count, memory mode (for TPCH), etc. PGO training phase is done with `--iterations 1` since for the training phase there is no need to perform multiple runs.

All PGO-related routines are done with [cargo-pgo](https://github.com/Kobzol/cargo-pgo).

All benchmarks are done on the same machine, with the same hardware/software during runs, with the same background "noise" (as much as I can guarantee, of course).

Some command-line commands for better understanding the process:

* Release benchmark (TPCH): `cargo run --release --bin tpch -- benchmark datafusion --iterations 5 --path ./data/tpch_sf1 -m --format parquet -o ./results/tpch_sf1_mem_release.json`
* PGO instrumentation phase (TPCH): `cargo pgo run -- --release --bin tpch -- benchmark datafusion --iterations 1 --path ./data/tpch_sf1 -m --format parquet -o ./results/tpch_sf1_mem_instrumented.json`
* PGO optimized benchmark: `cargo pgo optimize run -- --release --bin tpch -- benchmark datafusion --iterations 5 --path ./data/tpch_sf1 -m --format parquet -o ./results/tpch_sf1_mem_optimized.json`

Exactly the same process was used for any other benchmarks.

## Results

As a result, I will post several artifacts - raw benchmark results and a comparison made with [compare.py](https://github.com/apache/arrow-datafusion/blob/main/benchmarks/compare.py) script.

Results (comparison Release (left) vs PGO-optimized (right)):

* ClickBench: https://gist.github.com/zamazan4ik/968905023d99f1b4e14879467cce270d
* ClickBench partitioned: https://gist.github.com/zamazan4ik/207a0dd96c36036514b01f7639151877
* ClickBench extended: https://gist.github.com/zamazan4ik/c7c921833560715f1d9fa8ff0197e709
* Parquet: https://gist.github.com/zamazan4ik/27129b3b48658455cad8950cccfe4107
* Sort: https://gist.github.com/zamazan4ik/256a8909421b73a463704eceabef27a9
* TPCH Scale Factor 1: https://gist.github.com/zamazan4ik/3da01649d29a0dbf9747f4c3056809bd
* TPCH Scale Factor 1 in memory: https://gist.github.com/zamazan4ik/d69c23d5d7f42a056f7536f2d7f15d75
* TPCH Scale Factor 10: https://gist.github.com/zamazan4ik/6f93317d65161cc184eab15c5c24ffe2
* TPCH Scale Factor 10 in memory: https://gist.github.com/zamazan4ik/91df829cea91cb8c0efc5b5f64c49ebd

Raw results are attached as a `results.zip` archive (Raw results for Release, PGO-instrumented and PGO-optimized measurements): [results.zip](https://github.com/apache/arrow-datafusion/files/14538211/results.zip)

At least in the provided by project benchmarks, there are measurable improvements in many cases.

Maybe mentioning these results somewhere in the README (or any other user-facing documentation) can be a good idea. Also, I think, it will be worth performing benchmarks in more scenarios.

Please do not treat the issue as a bug or something like that - it's just a benchmark report.

**Comment by Dandandan** [comment]
Thanks for doing this @zamazan4ik cool stuff.
One thing I would be interested in is how does PGO generalize on one benchmark to another one, e.g. what does training on clickbench and running on tpch do and vice versa?

**Reply by zamazan4ik** [reply]
I of course can perform such benchmarks - I knew that you would ask such a question, and exactly for this reason I saved all gathered PGO profiles ;)

But I want to highlight that in real life it's not required to "generalize" over multiple benchmarks with PGO. What you really want is to optimize for your specific workload. And if TPCH differs *a lot* from Clickbench - PGO's positive effects will be reduced.

So before performing such benchmarks, I wanna ask you - why are you interested in such an experiment? Is there a practical reason for that or just theoretical knowledge? Maybe you have in mind some problem that requires such knowledge?

**Reply by zamazan4ik** [reply]
Here are the results:

* Clickbench benchmark, "Optimized with ClickBench profile" vs "Optimized with TPCH profile":  https://gist.github.com/zamazan4ik/7e904d3bd244335ee2c0d899747af488
* TPCH Scale Factor 10 benchmark, "Optimized with TPCH Scale Factor 10 profile" vs "Optimized with ClickBench profile": https://gist.github.com/zamazan4ik/bbbbf5e003d05ecbfb27056ed87168bd

With the Clickbench benchmark, we have an interesting thing to consider - with the TPCH profile it works faster! I guess it goes somehow from Query 29 since it slowed down. But it's just a guess - for better understanding profiling needs to be performed (e.g. compare flamegraphs for both cases and try to find the difference).

**Comment by alamb** [comment]
Thank you @zamazan4ik -- I agree this is really cool. I wonder what we should do with this excellent information? Some options I can think of:
1. Add a section to the documentation explaining that PGO can help up substantially (25%) and maybe offer some tips for users to use it?
2. Provide pre-gathered PGO data somehow, so users could build DataFusion with profiles guided from TPCH (or clickbench). 

In general I don't think many organizations will setup PGO with their own workload (as they often don't have an easy to run / representative benchmark available during builds or perhaps don't want to slow down their build process by running benchmarks at the same time)

**Reply by zamazan4ik** [reply]
> Add a section to the documentation explaining that PGO can help up substantially (25%) and maybe offer some tips for users to use it?

Yes, it would be a great option. It requires almost no resources to maintain (write once and link to this discussion for the results). In this case, users who are interested in optimizing `arrow-datafusion` more will be able to use this information as an additional optimization opportunity. I have several examples of how such documentation can be written (it's for applications but anyway - for a library case it should look a similar way):

* ClickHouse: https://clickhouse.com/docs/en/operations/optimizing-performance/profile-guided-optimization
* Databend: https://databend.rs/doc/contributing/pgo
* Vector: https://vector.dev/docs/administration/tuning/pgo/
* Nebula: https://docs.nebula-graph.io/3.5.0/8.service-tuning/enable_autofdo_for_nebulagraph/
* GCC: Official [docs](https://gcc.gnu.org/install/build.html), section "Building with profile feedback" (even AutoFDO build is supported)
* Clang:
  - https://llvm.org/docs/HowToBuildWithPGO.html
  - https://llvm.org/docs/AdvancedBuilds.html
* Rustc: https://rustc-dev-guide.rust-lang.org/building/optimized-build.html#profile-guided-optimization
* tsv-utils: https://github.com/eBay/tsv-utils/blob/master/docs/BuildingWithLTO.md

> Provide pre-gathered PGO data somehow, so users could build DataFusion with profiles guided from TPCH (or clickbench).

Unfortunately, this way is a bit trickier in practice. Pre-gathered PGO profiles have multiple issues - e.g. incompatibilities between different compiler versions, a profile skew (when a PGO profile is gathered for an older version of the code. When time flies, pre-gathered PGO profiles become less and less efficient so some kind of regular PGO profile regeneration is required). 

I could suggest another similar way - integrate into the build scripts the way to build the library with enabled PGO (based on some workload like TPCH, Clickbench, any other target workload, or any combination of them - it's up to discussion). On the one hand, users will be able to build the PGO-optimized version of the library. On another hand, you won't waste your maintenance resources on maintaining always up-to-date pre-gathered PGO profiles (however, this process can be simplified with CI).

Some examples of PGO build integration into the build scripts:

* Rustc: a CI [tool](https://github.com/rust-lang/rust/tree/master/src/tools/opt-dist) for the multi-stage build
* GCC:
  - Official [docs](https://gcc.gnu.org/install/build.html), section "Building with profile feedback" (even AutoFDO build is supported)
  - A [part](https://github.com/gcc-mirror/gcc/blob/master/configure#L7896) in a "wonderful" `configure` script.
* Clang:
  - [Docs](https://llvm.org/docs/HowToBuildWithPGO.html)
  - [MinGW build script](https://github.com/msys2/MINGW-packages/commit/4dd91d1d4dfef17f1f451c3a8f59303be855e4b5)
* Python:
  - CPython: [README](https://github.com/python/cpython#profile-guided-optimization)
  - Pyston: [README](https://github.com/pyston/pyston#building)
* Go: [Bash script](https://github.com/golang/go/blob/master/src/cmd/compile/profile.sh)
* Swift: [CMake script](https://github.com/apple/swift/blob/main/CMakeLists.txt#L364)
* V8: [Bazel flag](https://github.com/v8/v8/blob/main/BUILD.gn#L184)
* ChakraCore: [Scripts](https://github.com/chakra-core/ChakraCore/tree/master/Build/scripts/pgo)
* Chromium: [Script](https://chromium.googlesource.com/chromium/src/build/config/+/refs/heads/main/compiler/pgo/BUILD.gn)
* Firefox: [Docs](https://firefox-source-docs.mozilla.org/build/buildsystem/pgo.html)
   - Thunderbird has PGO support too
* PHP - [Makefile command](https://github.com/php/php-src/blob/master/build/Makefile.global#L138) and old Centminmod [scripts](https://github.com/centminmod/php_pgo_training_scripts)
* MySQL: [CMake script](https://github.com/mysql/mysql-server/blob/8.0/cmake/fprofile.cmake)
* YugabyteDB: [GitHub commit](https://github.com/yugabyte/yugabyte-db/commit/34cb791ed9d3d5f8ae9a9b9e9181a46485e1981d)
* FoundationDB: [Script](https://github.com/apple/foundationdb/blob/1a6114a66f3de508c0cf0a45f72f3687ba05750c/contrib/generate_profile.sh)
* Zstd: [Makefile](https://github.com/facebook/zstd/blob/dev/programs/Makefile#L232)
* [Foot](https://codeberg.org/dnkl/foot): [Scripts](https://codeberg.org/dnkl/foot/src/branch/master/pgo)
* Windows Terminal: [GitHub PR](https://github.com/microsoft/terminal/pull/10071)
* Pydantic-core: [GitHub PR](https://github.com/pydantic/pydantic-core/pull/741)
* file.d: [GitHub PR](https://github.com/ozontech/file.d/pull/469)
* OceanBase: [CMake flag](https://github.com/oceanbase/oceanbase/blob/master/cmake/Env.cmake#L55)
* ISPC: [CMake scipts](https://github.com/ispc/ispc/tree/main/superbuild)
* NodeJS: [Configure script](https://github.com/nodejs/node/commit/9be15559cc0bfe506d9cdfba4ad0f4beacf5ce17)
* Android Open Source Project (AOSP):
  - [Official documentation](https://source.android.com/docs/core/perf/pgo)
  - Committed PGO profiles: [repository](https://android.googlesource.com/toolchain/pgo-profiles/+/refs/heads/main)
* DMD: [Custom build rule](https://github.com/dlang/dmd/blob/master/compiler/src/build.d#L553)
* LDC: [GitHub action](https://github.com/ldc-developers/ldc/blob/master/.github/actions/2a-build-pgo/action.yml)
* tsv-utils: [Makefile](https://github.com/eBay/tsv-utils/blob/master/makefile#L56)
* Erlang OTP: [Makefile](https://github.com/erlang/otp/blob/master/Makefile.in#L484)
* Clingo (PGO enabled only in Spack): [Package recipe](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/clingo-bootstrap/package.py#L96)
* SWI-Prolog:
  - [Script](https://github.com/SWI-Prolog/swipl-devel/blob/master/scripts/pgo-compile.sh)
  - [CMake module](https://github.com/SWI-Prolog/swipl-devel/blob/master/cmake/PGO.cmake)
* hck: [Justfile](https://github.com/sstadick/hck/blob/master/justfile#L27)

If you have some prebuilt versions of the library (e.g. a Python wheel), you can think about pre-optimizing these prebuilt binaries with PGO (based on TPCH, Clickbench, etc.). As an example - Pydantic-core: [GitHub PR](https://github.com/pydantic/pydantic-core/pull/741).

> In general I don't think many organizations will set up PGO with their workload (as they often don't have an easy-to-run / representative benchmark available during builds or perhaps don't want to slow down their build process by running benchmarks at the same time)

It's a pity but I agree with you. Current PGO adoption across the industry is low (except for companies like Google and Facebook - which use PGO and similar optimization technologies like LLVM BOLT). This is the situation that I am trying to change by showing positive PGO effects for different applications.

**Reply by alamb** [reply]
Thank you very much @zamazan4ik  -- I have filed https://github.com/apache/arrow-datafusion/issues/9561 to add documentation
